[wasm-mt] Use asset loading for dotnet.worker.js; update WasmAppBuilder

lambdageek opened the pull request:
Enables using asset loading to get the `dotnet.worker.js` file that provides the emscripten pthread worker code.
Also allows specifying the number of pre-allocated workers that will be created at startup using MSBuild properties.

Fixes https://github.com/dotnet/runtime/issues/68509 and fixes https://github.com/dotnet/runtime/issues/68397 and fixes https://github.com/dotnet/runtime/issues/72606

- Override Emscripten `PThread.allocateUnusedWorker`
   We want to use our own allocateUnusedWorker because we want to load
   `dotnet.worker.js` using our asset loading machinery.

   Unfortunately, Emscripten first calls allocateUnusedWorker very early (from
   `PThread.init`) to pre-allocate the pthread worker pool.

   So we set Emscripten's own pthread worker pool to size 0 and make our own.  This requires calling `loadWasmModuleToWorker` during our startup because Emscripten deletes their code that normally does it (in "receiveInstance" in "createWasm" in "emscripten/src/preamble.js") when the pthread pool size is 0.

   Also added a pthreadPoolSize field to MonoConfig to allow specifying the initial pthread pool size in mono-config.json

- Add `IncludeThreadsWorker` and `PThreadPoolSize` props to WasmAppBuilder

   `IncludeThreadsWorker` adds the `"js-module-threads"` asset to the `mono-config.json`

   `PThreadPoolSize` can be -1 or >=0 to specify the number of workers that will be pre-allocated at startup for the pthread worker pool.  -1 means use the default compiled into `dotnet.js`

- Reorganize the pthreads TS code by moving `Internals` (access API that digs through Emscripten's pthreads implementation) to its own module. And add types.

- Replace emscripten's `allocateUnusedWorker` function with our own that goes through the asset loading API.

- Update samples

- Set up console proxying for the workers.
   This is done by sending a message from the main thread to the pthread workers with the current `MonoConfig` on our
    dedicated channel.  (This means the proxying is setup asynchronously, so if the worker is busy before it receives the message, it may not start redirecting messages right away).

lambdageek committed:
[wasm-mt] Override Emscripten PThread.allocateUnusedWorker

We want to use our own allocateUnusedWorker because we want to load
`dotnet.worker.js` using our asset loading machinery.

Unfortunately, Emscripten first calls allocateUnusedWorker very early (from
`PThread.init`) to pre-allocate the pthread worker pool.

So we set Emscripten's own pthread worker pool to size 0 and make our
own.  This requires calling `loadWasmModuleToWorker` during our
startup because Emscripten deletes their code that normally does
it (in "receiveInstance" in "createWasm" in
"emscripten/src/preamble.js") when the pthread pool size is 0.

Also added a pthreadPoolSize field to MonoConfig to allow specifying
the initial pthread pool size in mono-config.json

lambdageek committed:
Add IncludeThreadsWorker and PThreadPoolSize props to WasmAppBuilder

IncludeThreadsWorker adds the js-module-threads asset to the
mono-config

PThreadPoolSize can be -1 or >=0 to specify the number of workers that
will be pre-allocated at startup for the pthread worker pool.  -1
means use the default compiled into dotnet.js

lambdageek committed:
Move emscripten PThread internals access to a separate module

and add types

lambdageek committed:
Load js-module-threads asset in replacement allocateUnusedWorker

lambdageek committed:
Update samples to explicitly enable threading / perftracing

Makes the WasmAppBuilder include the threads worker module

lambdageek committed:
tighten up Internals types

lambdageek commented:
/azp run runtime-wasm

pavelsavara commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -149,6 +153,9 @@ async function preRunAsync(userPreRun: (() => void)[]) {
     await afterInstantiateWasm.promise;
     await afterPreInit.promise;
     if (runtimeHelpers.diagnosticTracing) console.debug("MONO_WASM: preRunAsync");
+    if (MonoWasmThreads) {
```
this need be after `mono_wasm_load_config` so that you could get the config for `pthreadPoolSize`

pavelsavara replied:
maybe I'm confused. Sorry.

lambdageek replied:
I verified that if I set `<_WasmPThreadPoolSize>6</_WasmPThreadPoolSize>` in the `browser-threading` sample project, I see this in the console 6 times: `MONO_WASM: replacementAllocateUnusedWorker` and the DevTools debugger shows the right number of workers.  So at least for the non-Blazor path it's working

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-mt-assets

lambdageek commented:
/azp run runtime-wasm

lambdageek commented:
@pavelsavara @radical any more feedback?  Anything for WasmApp.targets or the Task?

maraf commented on src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts:
```diff
@@ -0,0 +1,43 @@
+
+/// We replace Module["PThreads"].allocateUnusedWorker with this version that knows about assets
+function replacementAllocateUnusedWorker(): void {
+    console.debug("MONO_WASM: replacementAllocateUnusedWorker");
```
```suggestion
    if (runtimeHelpers.diagnosticTracing)
        console.debug("MONO_WASM: replacementAllocateUnusedWorker");
```

lambdageek committed:
apply review feedback

Co-authored-by: Marek Fišera <mara@neptuo.com>

lambdageek committed:
fix import

radical commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -314,6 +316,8 @@ private bool ExecuteInternal ()
         config.Assets.Add(new WasmEntry ("dotnet.wasm") );
         config.Assets.Add(new CryptoWorkerEntry ("dotnet-crypto-worker.js") );
+        if (IncludeThreadsWorker)
+                config.Assets.Add(new ThreadsWorkerEntry ("dotnet.worker.js") );
```
```suggestion
            config.Assets.Add(new ThreadsWorkerEntry ("dotnet.worker.js") );
```

radical commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -322,6 +326,11 @@ private bool ExecuteInternal ()
                     config.RemoteSources.Add(source.ItemSpec);
         }
 
+        if (PThreadPoolSize < -1) {
+            throw new LogAsErrorException($"PThreadPoolSize must be -1, 0 or positive, but got {PThreadPoolSize}");
+        } else
+            config.Extra["pthreadPoolSize"] = PThreadPoolSize;
```
nit:
```suggestion
        if (PThreadPoolSize < -1)
        {
            throw new LogAsErrorException($"PThreadPoolSize must be -1, 0 or positive, but got {PThreadPoolSize}");
        }
        else
        {
            config.Extra["pthreadPoolSize"] = PThreadPoolSize;
        }
```

radical reviewed (approved):
I'm not very familiar with `src/mono/wasm/runtime`, so I can't much review those changes. Rest of it looks fine. I would suggest running `runtime-wasm` though, and if the tests pass then it should be good to merge 👍

radical commented:
/azp run runtime-wasm

pavelsavara commented on src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts:
```diff
@@ -0,0 +1,44 @@
+    const asset = resolve_asset_path("js-module-threads");
+    const uri = asset.resolvedUrl;
+    mono_assert(uri !== undefined, "could not resolve the uri for the js-module-threads asset");
+    const worker = new Worker(uri);
```
we should subscribe onError here

pavelsavara replied:
and maybe proxy the console in the worker, similar way how it was done in crypto worker. See how we passed the json.stringify(runtimeHelpers.config) and then only forward console when enabled.

lambdageek replied:
> we should subscribe onError here

Emscripten installs an onerror handler of their own that will overwrite ours.  we can set one up later (in `afterLoadWasmModuleToWorker`) but we should probably not use the worker onerror handler for anything that we want to handle - we should catch exceptions around  our code and post it using our custom message channel.  We have a task to do that in the tracking issue.

lambdageek replied:
> and maybe proxy the console in the worker, similar way how it was done in crypto worker. See how we passed the json.stringify(runtimeHelpers.config) and then only forward console when enabled.

ok, i'll see if I can add that over the weekend; if not, there's a task to do it in the tracking issue

pavelsavara commented on src/mono/wasm/wasm.proj:
```diff
@@ -125,7 +125,7 @@
       <_EmccCommonFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s USE_PTHREADS=1" />
       <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-Wno-pthreads-mem-growth" />
-      <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s PTHREAD_POOL_SIZE=4" />
+      <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s PTHREAD_POOL_SIZE=0" />
```
i don't know if this is the case when the flag would be written into .rsp file in workload and could not be switched back on customer machine. just asking. And maybe it's OK

lambdageek replied:
Yea if they use wasm-tools, they will inherit this flag. I'm not sure if it's possible to override it.

pavelsavara commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -28,6 +28,7 @@ import { instantiate_wasm_asset, mono_download_assets, resolve_asset_path, start
 import { BINDING, MONO } from "./net6-legacy/imports";
 import { readSymbolMapFile } from "./logging";
 import { mono_wasm_init_diagnostics } from "./diagnostics";
+import { preAllocatePThreadWorkerPool, instantiateWasmPThreadWorkerPool } from "./pthreads/browser";
```
I think you checked this before, but please check again that non-MT build doesn't have the code included (in release build).

lambdageek replied:
Yep, I verified that this is gone

pavelsavara reviewed (approved):
lgtm

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

lambdageek committed:
proxy pthread worker messages to websocket, if enabled

use a new MonoThreadMessageApplyMonoConfig message to send
the MonoConfig from the main thread to each worker when the workers
set up the communication channel to the main thread.

then if the diagnosticTracing property is true, redirect the worker
console logging to a websocket.

Fixes https://github.com/dotnet/runtime/issues/72606

lambdageek commented:
Added console proxying.  verified that it gets set up in the `browser-threading` sample and the worker messages are going through the proxy.  Did not try running via xharness, though, but i assume that part works

lambdageek commented:
/azp run runtime-wasm

lambdageek commented:
Failures seem to be https://github.com/dotnet/runtime/issues/73901.
